Updated HTTP2 Reader to fix missing header state

isaacrivriv opened the pull request:
Motivation:
When a malformed request comes in where a stream error is thrown, data is still left on the reader and not properly cleaned. When a valid request comes in, it is not properly parsed because of the leftover data being included.

Modifications:
Updated DefaultHttp2Reader to clean up leftover data if these types of stream exceptions come in the wire.

Result:
After this change, data should be properly parsed after a malformed request
Fixes #13788 

If there is no issue then describe the changes introduced by this PR.

hyperxpro commented:
Can you include test case?

normanmaurer commented:
@idelpivnitskiy @bryce-anderson PTAL

isaacrivriv commented:
Added a simple test case that fails before the changes in the PR

isaacrivriv commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java:
```diff
@@ -164,6 +166,11 @@ public void readFrame(ChannelHandlerContext ctx, ByteBuf input, Http2FrameListen
+            if (resetReader) {
+                resetReader = false;
+                readingHeaders = true;
+                input.skipBytes(input.readableBytes());
```
I think you're right, if there's more data already buffered it'll skip all of it. I'm trying to think of an alternative but if the length of the payload doesn't match the actual payload not sure how we can skip the correct number of bytes to read the next frame correctly.

bryce-anderson replied:
Instead of using a boolean flag we could instead do something like passing `input.readSlice(payloadLength)` into the `processPayloadState` method. That would require us to move the 
```
if (input.readableBytes() < payloadLength) {
    // Wait until the entire payload has been read.
     return;
}
```
check up out of the `processPayloadState` method to right before we slice and call it but tbh that feels more readable anyway.

Doing that has two nice properties and one negative: it's nice because it 'slices' exactly one frame from `input` which (1) makes it hard for `processPayloadState` to corrupt other frames and (2) makes consumption of the frame from `input` happen eagerly so you don't have to worry about skipping bytes on exception. Downside is that it does cost an extra allocation. On balance, I'd happily pay the allocation of the slice wrapper: it isn't a drop in the ocean imo.

isaacrivriv replied:
That makes sense to me, but I think we'd hit some issues if the advertised `payloadLength` of the frame doesn't match the actual payload. In the use case I'm looking at, I see two cases where we could see issues
- __Advertised Payload Length greater than payload__: More data skipped than actually included and next frame would not be properly read.
- __Advertised Payload Length less than payload__: Less data skipped than actually included and data will be left over and next frame would not be properly read.

bryce-anderson replied:
It's convenient to think of the H2 protocol as having (at least) two layers: the framing layer (the first 9 octets of a frame) and then the actual frame payloads themselves which will have some consistency requirements. What I'm suggesting leverages the fact that the length from the framing layer should be authoritative in terms of frame boundaries.
What to do if we detect corruption/inconsistencies in the payload (which could be that the payload is short or long, for example) will be case specific and gets into the details of what @idelpivnitskiy pointed to in his suggestions. This is actually a third advantage of what I'm suggestion: extra bytes can signal corruption and if we slice we can detect that. For example, a RST_STREAM frame with size other than 4 octets should be a connection error (end of [6.4](https://datatracker.ietf.org/doc/html/rfc9113#section-6.4)). However in main if we received a RST_STREAM frame of size 5 (as determined by the frame header) where the last byte is meaningless instead of sending a connection error and potentially alerting the peer to a bug in their implementation we'll instead end up with a corrupt framing layer: we won't fail because we don't enforce the size and we won't consume that extra byte from the data stream so it will end up being interpreted as the first byte of the next frames header.

isaacrivriv replied:
Yeah that's exactly the issue I was seeing, leftover data is parsed as part of the next frame when malformed frames like that are parsed. So with @idelpivnitskiy comments and this thread I'm thinking we can implement the slice mechanism but there needs to be more checks implemented so that the payload length matches the actual payload and for those cases where we don't find a match we can alert by sending a Go Away frame?

bryce-anderson replied:
That sounds right: whether to send a GOAWAY or a RST_STREAM is frame type specific and described (usually at the ends) of the frame definitions. I don't think you have to implement all those checks as part of this PR: they could be done as a follow up. Finding this and making sure we keep a correct frame layer is already a huge step forward from where we are now.

bryce-anderson replied:
Taking a closer look it seems most of them (if not all) already verify the frame length and do the right thing wrt RST vs GOAWAY but its done as part of the `processHeaderState` state so we haven't been able to slice before the exception is thrown, and we may not even have that many bytes available, so we'd need to move the verification to a new method and perform it right before we `processPayloadState(ctx, input.slice(..), ..)`.

idelpivnitskiy commented on codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2FrameReaderTest.java:
```diff
@@ -386,6 +386,40 @@ public void execute() throws Throwable {
+                    }
+                }
+            });
+            // Verify that after stream exception we accept new stream requests
```
@isaacrivriv I read description of #13788 to understand expected behavior, but it seems like this is not what RFC wants. [Section 4.2](https://datatracker.ietf.org/doc/html/rfc9113#section-4.2) says:

> An endpoint MUST send an error code of [FRAME_SIZE_ERROR](https://datatracker.ietf.org/doc/html/rfc9113#FRAME_SIZE_ERROR) if a frame exceeds the size defined in [SETTINGS_MAX_FRAME_SIZE](https://datatracker.ietf.org/doc/html/rfc9113#SETTINGS_MAX_FRAME_SIZE), exceeds any limit defined for the frame type, or is too small to contain mandatory frame data. A frame size error in a frame that could alter the state of the entire connection MUST be treated as a [connection error](https://datatracker.ietf.org/doc/html/rfc9113#ConnectionErrorHandler) ([Section 5.4.1](https://datatracker.ietf.org/doc/html/rfc9113#ConnectionErrorHandler))


Then section 5.4.1 describes that the state of the entire connection is expected to be corrupted at that point and we must close it. It's not expected that we could recover and process more frames on the same connection.

If we don't already have one, it will be helpful if you could implement an integration test with a full HTTP/2 server and a faked client that send a malformed frame, then assert that the server responds with `FRAME_SIZE_ERROR`, then closes connection.

isaacrivriv replied:
In this case, the behavior I'm seeing with the response from a Netty HTTP2 server matches the RFC specification where in some cases even though a malformed frame comes in, it doesn't actually shut down the connection but just resets a stream and more streams can be accepted after that. [Section 6.3](https://datatracker.ietf.org/doc/html/rfc9113#name-priority) is an example

> A PRIORITY frame with a length other than 5 octets MUST be treated as a [stream error](https://datatracker.ietf.org/doc/html/rfc9113#StreamErrorHandler) ([Section 5.4.2](https://datatracker.ietf.org/doc/html/rfc9113#StreamErrorHandler)) of type [FRAME_SIZE_ERROR](https://datatracker.ietf.org/doc/html/rfc9113#FRAME_SIZE_ERROR).

As it is currently working if I send a priority frame with length 4, I correctly get a reset frame back with a frame size error (which is what the RFC states) but starting another stream with a valid request after that, it is not parsed correctly due to leftover data from the previous request.

idelpivnitskiy replied:
The most important part is: 

> A frame size error in a frame that could alter the state of the entire connection MUST be treated as a [connection error](https://datatracker.ietf.org/doc/html/rfc9113#ConnectionErrorHandler) ([Section 5.4.1](https://datatracker.ietf.org/doc/html/rfc9113#ConnectionErrorHandler))

It's ok to send `RST_STREAM` frame first. But then, how can we trust the frame if it has incorrect format? If we can not trust it, we don't know how many bytes to skip. It could happen that we received total garbage that just happen to start with "priority frame header". So, if we don't know how to continue, we should send GO_AWAY and close connection.

isaacrivriv replied:
The RFC states the following after that statement which confuses me

> A frame size error in a frame that could alter the state of the entire connection MUST be treated as a [connection error](https://datatracker.ietf.org/doc/html/rfc9113#ConnectionErrorHandler) ([Section 5.4.1](https://datatracker.ietf.org/doc/html/rfc9113#ConnectionErrorHandler)); **this includes any frame carrying a [field block](https://datatracker.ietf.org/doc/html/rfc9113#FieldBlock) ([Section 4.3](https://datatracker.ietf.org/doc/html/rfc9113#FieldBlock)) (that is, [HEADERS](https://datatracker.ietf.org/doc/html/rfc9113#HEADERS), [PUSH_PROMISE](https://datatracker.ietf.org/doc/html/rfc9113#PUSH_PROMISE), and [CONTINUATION](https://datatracker.ietf.org/doc/html/rfc9113#CONTINUATION)), a [SETTINGS](https://datatracker.ietf.org/doc/html/rfc9113#SETTINGS) frame, and any frame with a stream identifier of 0.**

So in my head the RFC states the specifics of what "altering the state of the connection" is in that sentence which this particular case doesn't seem to apply. By sending a priority frame to a specific stream (not 0) we can hit this issue and doesn't really fall under those categories IMO. 

I think we should skip the bytes if the payload length matches the actual frame payload and send a reset stream which matches the RFC and continue reading for the cases where a stream reset is required. I also agree closing the connection on cases where the payload length doesn't actually match the frame payload considering that we are receiving garbage in those cases but I'm making sure this isn't against the RFC since it states otherwise about keeping the connection active.

The other question would be, how would we know if the frame that came in actually matches the payload length? By the time Netty start's parsing H2 data we could have more messages queued in the buffer which we're not aware of. I'm thinking maybe a check on every decode to peek at the payload length and compare to the actual frame length? I'm not aware of a way to cleanly distinguish frames from each other TBH.

bryce-anderson replied:
See my comment above but the TL;DR is that the length field from the frame header is authoritative on size of a frames payload. What to do if that payload size is too short or too long when decoding the frame is frame type specific and is described as part of the individual frame definitions.

bryce-anderson replied:
@normanmaurer, I am. This should prevent frame stream corruption for problems that result in a stream error at the validation phase.

One behavior change that I think is worth point out is that we'll now aggregate the data for the next frame before doing any of the per-frame validations other than max frame length. I think this is safe because we've already agreed to accept frames that big such as data and headers frames but it's worth a thought.

idelpivnitskiy replied:
@isaacrivriv thanks for pointing to the continuation of that sentence. That makes sense, for the current case we don't need to close the entire connection. Great catch!

bryce-anderson commented:
@isaacrivriv, I imagine you're busy but I wanted to see what your plan was for this. It's a great find and I'd love to see it fixed. 🙂

isaacrivriv commented:
Sorry for the late reply, I was trying to get to this last week but couldn't get it to a state I was happy with. I added a push with the changes we discussed in mind. I separated this following these steps
- Parsing the framing layer component of the frame (everything all the frame types have in common)
- Verifying the data read from the framing layer based off the different types
- Slicing the buffer after verifying and processing the payload for this specific frame
- Resetting the flags specifically the `readingHeaders` flag to the default value once a stream reset occurs while processing the headers/payload

The test I added passes and ran all the tests locally for http2 which pass as well. I did some manual testing to replicate garbage requests submitting more data than actually announced in the payloadLength and that still fails as expected. I'm not sure of a way to be able to identify these cases and respond to the client that something may be wrong with their implementation if anyone had anything in mind.

bryce-anderson commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java:
```diff
@@ -229,28 +235,18 @@ private void processHeaderState(ByteBuf in) throws Http2Exception {
         switch (frameType) {
             case DATA:
-                readDataFrame(ctx, in, payloadEndIndex, listener);
+                readDataFrame(ctx, in, payloadLength, listener);
```
All of these methods still work despite passing in a value with different meaning only because the writer index now matches the payload length. Now that you've sliced the buffer really we don't need to pass in the end length as each method could derive it from the passed in slice itself.

isaacrivriv replied:
True, I'll update the method signatures to make this cleaner.

bryce-anderson commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java:
```diff
@@ -496,10 +493,10 @@ private void readPriorityFrame(ChannelHandlerContext ctx, ByteBuf payload,
         long word1 = payload.readUnsignedInt();
         boolean exclusive = (word1 & 0x80000000L) != 0;
         int streamDependency = (int) (word1 & 0x7FFFFFFFL);
+        short weight = (short) (payload.readUnsignedByte() + 1);
         if (streamDependency == streamId) {
             throw streamError(streamId, PROTOCOL_ERROR, "A stream cannot depend on itself.");
         }
-        short weight = (short) (payload.readUnsignedByte() + 1);
```
fwiw, now that you've sliced the buffer I don't think you need to do this re-ordering.

bryce-anderson commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java:
```diff
@@ -143,27 +143,30 @@ public void readFrame(ChannelHandlerContext ctx, ByteBuf input, Http2FrameListen
             } while (input.isReadable());
         } catch (Http2Exception e) {
             readError = !Http2Exception.isStreamError(e);
+            // Stream error found, reset to reading headers once again for next request coming in
```
```suggestion
            // Stream error found, reset to reading headers once again for next frame coming in
```

bryce-anderson commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java:
```diff
@@ -143,27 +143,30 @@ public void readFrame(ChannelHandlerContext ctx, ByteBuf input, Http2FrameListen
+                verifyHeaderState();
                 // The header is complete, fall into the next case to process the payload.
                 // This is to ensure the proper handling of zero-length payloads. In this
                 // case, we don't want to loop around because there may be no more data
                 // available, causing us to exit the loop. Instead, we just want to perform
                 // the first pass at payload processing now.
-                processPayloadState(ctx, input, listener);
-                if (!readingHeaders) {
-                    // Wait until the entire payload has arrived.
+
+                // Wait until the entire payload has been read.
+                if (input.readableBytes() < payloadLength) {
                     return;
                 }
+                // Slice to work only on the frame being read
+                ByteBuf framePayload = input.readSlice(payloadLength);
+                processPayloadState(ctx, framePayload, listener);
```
Some of the verifications result in a stream error, not a connection error. If that happens I think we end up with a corrupt stream. To fix it we can simply move verification to after we've called input.readSlice(..) so that we've consumed the frame from the framing layer. We've already committed to reading frames this large so it doesn't seem like a potential attack vector. The only downside seems to be that for connection errors it can potentially delay closing the session.

isaacrivriv replied:
Agree, I forgot that we aren't skipping the bytes anymore like the original solution if this happens. I'll move this verification after slicing

bryce-anderson commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java:
```diff
@@ -229,28 +235,18 @@ private void processHeaderState(ByteBuf in) throws Http2Exception {
                 verifyUnknownFrame();
                 break;
         }
+        readingHeaders = false;
```
We should be able to simplify the management of this variable: we switch back to `readingHeaders = true` as soon as we consume the frame payload, so we can set it right after we `.readSlice(..)` and before we attempt to process the payload.

bryce-anderson reviewed (commented):
This is coming along really nicely. Thank you @isaacrivriv!

bryce-anderson commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java:
```diff
@@ -402,20 +392,20 @@ private void verifyUnknownFrame() throws Http2Exception {
         // Determine how much data there is to read by removing the trailing
         // padding.
-        int dataLength = lengthWithoutTrailingPadding(payloadEndIndex - payload.readerIndex(), padding);
+        int dataLength = lengthWithoutTrailingPadding(payloadLength - payload.readerIndex(), padding);
```
I think these can be even simpler: 
```suggestion
        int dataLength = lengthWithoutTrailingPadding(payload.readableBytes(), padding);
```

isaacrivriv replied:
Agree, I'll make the changes to all of these

bryce-anderson commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java:
```diff
@@ -402,20 +392,20 @@ private void verifyUnknownFrame() throws Http2Exception {
-        int dataLength = lengthWithoutTrailingPadding(payloadEndIndex - payload.readerIndex(), padding);
+        int dataLength = lengthWithoutTrailingPadding(payloadLength - payload.readerIndex(), padding);
 
         ByteBuf data = payload.readSlice(dataLength);
```
Do you think we could just adjust the writer index on `payload`?

isaacrivriv replied:
So instead of slicing the data buffer just move the writer index of the payload to dataLength?

bryce-anderson replied:
Yes: I think we sliced the payload previously because we didn't want to leak the parent buffer but since we've already sliced it there shouldn't be any harm in sharing that. At a minimum, I expect padding to be rare so if there isn't any padding to slice out we can certainly just send `payload` along.

And to be more precise, I think we should do `payload.writeIndex(payload.readerIndex() + dataLength)`. `payload.readerIndex()` should always be 0 now, but it will make the code more resilient to future changes.

isaacrivriv replied:
Ok cool I'll make that change and saw the payload being further sliced while reading a Go Away frame so I'm updating that as well

bryce-anderson reviewed (commented):
Just two small suggestions. I really like how this turned out.

isaacrivriv commented:
Thanks for the detailed review @bryce-anderson! I'm getting a better understanding on how everything works behind the scene while doing this. I pushed out changes based on the comments. I also had to update the test I added to add dummy data to the buffer for the reader to consume it and continue since the logic to identify if the request is valid or not (extra data or not enough data for the request) is not added. Let me know what you think!

bryce-anderson reviewed (approved):
Thanks @isaacrivriv!

normanmaurer commented:
/cc @ejona86

idelpivnitskiy commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java:
```diff
@@ -174,25 +175,26 @@ public void readFrame(ChannelHandlerContext ctx, ByteBuf input, Http2FrameListen
-
-        // Read the header and prepare the unmarshaller to read the frame.
         payloadLength = in.readUnsignedMedium();
+        frameType = in.readByte();
```
Consider adding a comment for future devs why we read the full header before throwing an excepting on case `payloadLength > maxFrameSize`

isaacrivriv replied:
I'm looking at this again and since it's a connectionError which closes the connection I don't think we have to read the full headers before throwing the exception so I'm undoing these changes

idelpivnitskiy reviewed (approved):
Few minor recommendations, otherwise LGTM:

isaacrivriv commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java:
```diff
@@ -174,25 +175,26 @@ public void readFrame(ChannelHandlerContext ctx, ByteBuf input, Http2FrameListen
+        return true;
+    }
 
+    private void verifyHeaderState() throws Http2Exception {
```
I agree, renaming it makes more sense I'll make this change

isaacrivriv commented on codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java:
```diff
@@ -233,24 +235,13 @@ private void processHeaderState(ByteBuf in) throws Http2Exception {
-        // We have consumed the data, next time we read we will be expecting to read a frame header.
-        readingHeaders = true;
-
         // Read the payload and fire the frame event to the listener.
```
Sounds good with me, I'll make this change

normanmaurer commented:
Let's plan to get this in after we cut the next release just to give it a bit more time to bake and for people like @ejona86 to check it out.

normanmaurer commented:
@isaacrivriv thanks a lot! I also cherry-picked into main
